src/core/server/http/types.ts

```diff
@@ -264,7 +266,10 @@ export interface InternalHttpServiceSetup
   extends Omit<HttpServiceSetup, 'createRouter' | 'registerRouteHandlerContext'> {
   auth: HttpServerSetup['auth'];
   server: HttpServerSetup['server'];
-  createRouter: (path: string, plugin?: PluginOpaqueId) => IRouter;
+  createRouter: <Context extends RequestHandlerContext = RequestHandlerContext>(
+    path: string,
+    plugin?: PluginOpaqueId
+  ) => IRouter<Context>;
   getAuthHeaders: GetAuthHeaders;
   registerRouteHandlerContext: <T extends keyof RequestHandlerContext>(
     pluginOpaqueId: PluginOpaqueId,
```

Comment:
Now that we are more or less deprecating the
```ts
declare module 'src/core/server' {
  interface RouteHandlerContext {
    myField: MyType;
  }
}
```
approach in favor of that, I think the 
```ts
  registerRouteHandlerContext: <T extends keyof RequestHandlerContext>(
    contextName: T,
    provider: RequestHandlerContextProvider<T>
  ) => RequestHandlerContextContainer;
```

should change to something like
```ts
  registerRouteHandlerContext: <T extends string>(
    contextName: T,
    provider: RequestHandlerContextProvider<T>
  ) => RequestHandlerContextContainer;
```
Else, plugin would still need to use the `declare` statement to be able to add the key to the `RequestHandlerContext` before registering. WDYT?

Reply:
Only other option I can is making the context type here generic as well, so it'd be:

```ts
  registerRouteHandlerContext: <
    TContext extends RequestHandlerContext,
    T extends keyof TContext
  >(
    contextName: T,
    // this would probably need to be changed to accept a type arg as well
    // ...and possibly IContextContainer would need to be changed?
    provider: RequestHandlerContextProvider<T>
  ) => RequestHandlerContextContainer;
```

Only concern with this is that the generics in IContextContainer are already confusing...adding more generics might make this worse?

Reply:
I think we have enough generics in this part of the code yes and would like to avoid introducing new confusing ones yes. Also with your proposition, I think the call to registerRouteHandlerContext will always to specify the `TContext` type as it cannot be guessed from the call:

```ts
http.registerRouteHandlerContext<MyCustomRequestHandlerContext>('myKey', provider)
```

Although, my example is also false:

```typescript
  registerRouteHandlerContext: <T extends string>(
    contextName: T,
    provider: RequestHandlerContextProvider<T>
  ) => RequestHandlerContextContainer;
```

Still doesn't work without the `declare` statement, as the provider is still expected to be a `RequestHandlerContextProvider<T>`, which is:

```typescript
export type RequestHandlerContextProvider<
  TContextName extends keyof RequestHandlerContext
> = IContextProvider<RequestHandler<any, any, any>, TContextName>;
```

which is(...)

```typescript
export type IContextProvider<
  THandler extends HandlerFunction<any>,
  TContextName extends keyof HandlerContextType<THandler>
> = (
  context: Partial<HandlerContextType<THandler>>,
  ...rest: HandlerParameters<THandler>
) =>
  | Promise<HandlerContextType<THandler>[TContextName]>
  | HandlerContextType<THandler>[TContextName];
```

Meaning it still expects to resolve it's type from a property of `RequestHandlerContext`.

Changing the `registerRouteHandlerContext` signature is going to be way harder than anticipated. Really not sure of the direction to take, the context types in `src/core/utils/context.ts` are really hard to follow...

Reply:
> My guess is that we should cut the relation between the context name and the context provider when registering, as there is no longer a direct link between the name of the context and the provider's result type.

I personally prefer that we maintain type safety here so that we can be sure that the provider that is registered is actually returning what it is supposed to.

However, I agree the generics in `src/core/utils/context.ts` are already confusing. One option could be to only add the generics to `registerRouteHandlerContext` and cast it to `any` when passing it over to `IContextContainer#registerContext`. Reason I think this may be acceptable at this time is that we may be able to replace the super generic `IContextContainer` with a concrete implementation just for route handlers once we remove ApplicationService's context completely (already deprecated).